[FEAT] Entity 초기 설계

twotwobread opened the pull request:
## PR Description 🗒️

## 추가/변경 사항 및 설명 📝
- Entity 초기 설계
  - 팀원들의 코드 리뷰를 이용하여 설계 수정 예정.
  - 코드 리뷰에 대한 답변을 실시간으로 주고 받아서 답변이 없는 것들이 존재.
- @twotwobread 가 생각하는 필요한 Entity
  - Member(회원)
    - memberId : Long, GeneratedValue (멤버 고유 ID)
    - memberName :  String, NotNull (멤버 이름)
    - image : Image, (프로필 이미지) (1 - 1)
    - ranking : Integer, NotNull (멤버 대결 포인트 순위) -> 빼낼 수 있을 듯.
    - winningPoint : Integer, default = 0 (대결 승리 포인트) -> 빼낼 수 있을 듯.
    - winningCount : Integer, default = 0 (대결 승리 횟수) -> 빼낼 수 있을 듯.
    - battles : List<Battle>, default로 자료구조 생성해두기 (멤버가 참여한 배틀들) (1 - N)
    - posts : List<Post>, default로 자료구조 생성해두기 (멤버가 공유한 음악 게시물들) (1 - N)
  - Image(이미지)
    - imageId : Long, GeneratedValue (이미지 고유 ID)
    - imageUrl : String, NotNull (이미지 URL)
    - member : Member, (이미지 소유한 멤버) (1 - 1)
  - Post(음악 공유 게시글)
    - postId : Long, GeneratedValue (게시글 고유 ID)
    - member : Member, (게시자) (N - 1)
    - music : Music, (공유하는 음악) (N - 1)
    - isPossibleBattle : Boolean, NotNull (게시글의 대결 가능 여부)
    - likeCount : Integer, NotNull, default = 0 (게시글 좋아요 개수) -> 빼낼 수 있을 듯.
  - Music(음악)
    - musicId : Long, GeneratedValue (음악 고유 ID)
    - singer : String, NotNull (가수 이름)
    - musicName : String, NotNull (음악 제목)
    - genre : String, NotNull (음악 장르)
    - musicUrl : String, NotNull (음악 재생 관련해서 저장해놓기)
    - posts : List<Post>, default로 자료구조 생성해두기 (해당 음악을 공유하고 있는 게시글들) (1 - N)
  - Battle(음악 대결 게시글)
    - battleId : Long, GeneratedValue (대결 고유 ID)
    - ownerPost : Post, (대결 신청을 받는 음악 공유 게시글) (N - 1)
    - oppositePost : Post, (대결을 신청하는 음악 공유 게시글) (N - 1)
    - ownerVoteCount : Integer, NotNull, default = 0
    - oppositeVoteCount : Integer, NotNull, default = 0

## To Reviewers 🙏
- 커밋 메시지와 TODO를 이용해서 헷갈리는 부분을 적었습니다!!
## Issue close ✂️
#30

twotwobread committed:
[DOCS] PR template 추가
- .github 하위에 pr template 추가.

close #1이수영_#1

twotwobread committed:
[CHORE] jacoco 설정을 위한 build.gradle 파일 수정.
- jacoco plugins 추가
- jacocoTestReport task 추가
- jacocoTestCoverageVerification task 추가
- test task 실행되고 추가한 두 task가 실행될 수 있게 finalizedBy 추가
close #이수영_#3

twotwobread committed:
[CHORE] 코드 커버리지에서 lombok 어노테이션을 제외하기 위한 설정파일 추가

close #이수영_#3

twotwobread committed:
[DEL] 해당 PR에 필요없는 부분 삭제

twotwobread committed:
[CHORE] 필요없는 주석 삭제

twotwobread committed:
Merge branch 'develop' into lsy

twotwobread committed:
Merge branch 'develop' into suyoung

twotwobread committed:
Merge branch 'develop' into suyoung

twotwobread committed:
[TEST] : 필요없는 테스트 삭제.

twotwobread committed:
[FEAT] Member 엔티티 추가.
- 대체 키를 이용하도록 구성.
- 랭킹, 포인트, 승수에 대한 부분 숫자로 표현
- 프로필 이미지를 위한 이미지 필드 추가.
- 멤버가 참여한 대결에 대한 정보를 위해 battles 추가.
- 멤버가 공유한 게시글에 대한 정보를 위해 posts 추가.

twotwobread committed:
[FEAT] Image 엔티티 추가.
- 대체 키 이용.
- 이미지 경로 String으로  추가.
- 해당 이미지를 프로필로 가진 멤버에 대한 부분 일대일 관계 설정.
- 멤버측에서 더 많은 작업이 발생할 것 같아서 외래키 주인을 멤버로 설정.

twotwobread committed:
[FEAT] Music 엔티티 추가.
- 대체키 이용.
- 가수 이름을 저장하는 필드 추가.
- 음악 제목을 저장하는 필드 추가.
- 장르를 enum 형식으로 저장.
- 음악 재생과 관련된 url을 저장.
- 해당 음악을 공유하는 게시글에 대한 정보를 일대다 관계로 설정.

- Music 엔티티가 있어야할까란 생각도 듬. 그 이유는 어짜피 Music을 이용한 기능이 없고 이렇게 되면 음악 검색 api를 이용하기 전에 존재하는 music인지부터 확인하는 작업이 필요해질것 같음.

twotwobread committed:
[FEAT] Post 엔티티 추가.
- 대체 키 이용.
- 대결 가능 여부에 대한 필드 추가.
- 좋아요 갯수를 숫자로 필드 추가.
- music에 대한 부분 다대일 관계로 설정.
- post에서 배틀이 필요한 이유는 배틀 쪽에서 Post의 정보가 필요해서 넣었음.

twotwobread committed:
[FEAT] Battle 엔티티 추가.
- 대체 키 이용
- 대결 신청을 받은 게시글의 득표 수에 대한 필드 추가.
- 대결을 신청한 게시글에 대한 득표 수 필드 추가.
- member가 필요한 이뉴는 member 측에서 자신이 참여한 battle의 정보를 가져오기 위함이다.
- battle에서는 두 개의 post 정보가 필요한데 각각 다른 필드로 구성되어도 될까?
  - 다르게 생각했던건 post를 List화 시킬 생각을 했는데 이것도 너무 이상하다.

twotwobread committed:
Merge pull request #31 from prgrms-web-devcourse/suyoung_#30

[FEAT] Entity 초기 설계

thguss commented on src/main/java/com/example/demo/model/Battle.java:
```diff
@@ -0,0 +1,36 @@
+@NoArgsConstructor
+public class Battle {
+
+	// TODO: Battle에서 두 개의 Post 정보를 가져야함. -> Post와의 관계가 다대일이 되는게 맞을지, Post가 다른 필드로 2번 오는게 맞을지
```
2개 지정으로 되어있으니 대결곡, 상대곡 2개 필드로 가져오는 것이 좋을 것 같아요~

park-se-jun replied:
battle 1 : post 2 의 관계가 좋을 것 같은 느낌쓰

thguss commented on src/main/java/com/example/demo/model/Battle.java:
```diff
@@ -0,0 +1,36 @@
+	Integer oppositeVoteCount;
+
+	@ManyToOne(fetch = FetchType.LAZY)
+	Member member;
```
member도 2개 필드가 필요한데, post에 포함되어 있으니 거기서 join으로 가져와도 될 것 같아요! (Battle 엔티티에서 빼고!)

thguss commented on src/main/java/com/example/demo/model/Battle.java:
```diff
@@ -0,0 +1,36 @@
+	// TODO: Battle에서 두 개의 Post 정보를 가져야함. -> Post와의 관계가 다대일이 되는게 맞을지, Post가 다른 필드로 2번 오는게 맞을지
+
+	@Id
+	@GeneratedValue
```
전략 논의해봅시다!

thguss commented on src/main/java/com/example/demo/model/Battle.java:
```diff
@@ -0,0 +1,36 @@
+	Integer ownerVoteCount;
+
+	@NotNull
+	Integer oppositeVoteCount;
```
따로 나눠놓은 거 좋네요~ 굳!!

- int가 아닌 Integer로 정의한 이유가 있을까요??
- 투표 수 계산 메소드도 생성하면 좋을 것 같아요~! + Vote Entity도 추가로 만들면 좋을 것 같아요!

thguss commented on src/main/java/com/example/demo/model/Image.java:
```diff
@@ -0,0 +1,25 @@
+
+@Entity
+@NoArgsConstructor
+public class Image {
```
어떤 목적의 이미지 엔티티인가요?? 유저 프로필 이미지라면 Member 엔티티에 imageUrl만 달아도 충분하다고 생각해요~!

park-se-jun replied:
이미지는 너무 추상적인것 같습니다. 프로필 이미지 인가요? 프로필 이미지면 따로테이블이 있는게 아니라 member 가 url을 들고 있는 형식 어떤가요?

thguss commented on src/main/java/com/example/demo/model/Member.java:
```diff
@@ -0,0 +1,42 @@
+@NoArgsConstructor
+public class Member {
+
+	// TODO: Member에서 Battles를 가지고 있는게 맞을 지 -> 직관적으론 가지고 있는게 맞다고 느껴지는데 Battle을 짜다보면 Battle에서 Member가 필요 없음.
```
유저별 참여한 Battle 리스트를 조회하는 것을 생각하면 가지고 있는 것이 좋다고 생각해요~
다만 Battle에서 Member가 필요 없다는 것에는 동의합니다. Member 엔티티에서만 Battle을 가지고 있어도 될 듯 해요!

park-se-jun replied:
그러면 member는 battle을 알지만 battle은 member를 몰라도 되는 단방향 관계는 어떤가요. 그리고 기획을 보면 battle이 user 정보를 필요로 하는 부분이 있었던 것 같아요(유저페이지 쪽)

thguss commented on src/main/java/com/example/demo/model/Member.java:
```diff
@@ -0,0 +1,42 @@
+public class Member {
+
+	// TODO: Member에서 Battles를 가지고 있는게 맞을 지 -> 직관적으론 가지고 있는게 맞다고 느껴지는데 Battle을 짜다보면 Battle에서 Member가 필요 없음.
+	// TODO: 포인트, 랭킹, 승수 이 부분이 애매함.
```
포인트, 랭킹, 승 수 필드를 만들고 계산 메소드도 별도로 필요하다고 생각됩니다.

thguss commented on src/main/java/com/example/demo/model/Member.java:
```diff
@@ -0,0 +1,42 @@
+
+	Integer ranking;
+	Integer winningPoint;
+	Integer winningCount;
```
이 또한 int가 아닌 Integer로 정의한 이유가 있을까요??

thguss commented on src/main/java/com/example/demo/model/Member.java:
```diff
@@ -0,0 +1,42 @@
+	Integer winningCount;
+
+	@OneToOne(fetch = FetchType.LAZY)
+	Image image;
```
String imageUrl로 충분하다고 생각됩니다~

thguss commented on src/main/java/com/example/demo/model/Member.java:
```diff
@@ -0,0 +1,42 @@
+	Long memberId;
+
+	@NotNull
+	String memberName;
```
API 설계 과정에서 url에 nickname을 사용했던 것으로 기억하는데, 통일시키면 좋을 것 같아요~

thguss commented on src/main/java/com/example/demo/model/Post.java:
```diff
@@ -0,0 +1,41 @@
+	Long postId;
+
+	@NotNull
+	Boolean isPossibleBattle;
```
boolean 대신 Boolean으로 하신 이유가 있을까요~?

thguss commented on src/main/java/com/example/demo/model/Battle.java:
```diff
@@ -0,0 +1,36 @@
+	Post ownerPost;
+
+	@ManyToOne(fetch = FetchType.LAZY)
+	Post oppositePost;
```
joinColumn 빠졌어용

thguss commented on src/main/java/com/example/demo/model/Battle.java:
```diff
@@ -0,0 +1,36 @@
+	Member member;
+
+	@ManyToOne(fetch = FetchType.LAZY)
+	Post ownerPost;
```
joinColumn이 빠진 것 같아요~!

thguss commented on src/main/java/com/example/demo/model/Post.java:
```diff
@@ -0,0 +1,41 @@
+	Member member;
+
+	@ManyToOne(fetch = FetchType.LAZY)
+	Music music;
```
joinColumn !

thguss commented on src/main/java/com/example/demo/model/Post.java:
```diff
@@ -0,0 +1,41 @@
+	Integer likeCount;
+
+	@ManyToOne(fetch = FetchType.LAZY)
+	Member member;
```
joinColumn이 빠진 것 같아요 :)

thguss replied:
Music 클래스로 바꾼다면 joinColumn 필요없을 것 같습니다 ㅎㅅㅎ

thguss commented on src/main/java/com/example/demo/model/Music.java:
```diff
@@ -0,0 +1,41 @@
+public class Music {
+
+	// TODO : Music을 따로 엔티티로 두는 것이 맞을지 -> 이렇게 되면 음악 검색 API 하기 이전에 쿼리를 던져야함.
+	// TODO : Music 엔티티에서 musicUrl을 따로 가지고 있는게 맞을지 -> 안 가지고 있으면 매번 API를 통해 음악 재생해줘야함.
```
Music을 엔티티로 두기 보다 클래스로 두는 것은 어떤가요?? JPA 스터디에서 사용한 Address 같은 Embedded 타입으로요!

thguss commented on src/main/java/com/example/demo/model/Music.java:
```diff
@@ -0,0 +1,41 @@
+
+	@Id
+	@GeneratedValue
+	Long musicId;
```
임베디드로 간다면 id는 필요없게 될 것 같아요~!

thguss commented on src/main/java/com/example/demo/model/Post.java:
```diff
@@ -0,0 +1,41 @@
+	Boolean isPossibleBattle;
+
+	@NotNull
+	Integer likeCount;
```
Like Entity가 추가되면 좋을 것 같아요!

park-se-jun commented on src/main/java/com/example/demo/model/Battle.java:
```diff
@@ -0,0 +1,36 @@
+	@NotNull
+	Integer oppositeVoteCount;
+
+	@ManyToOne(fetch = FetchType.LAZY)
```
memeber 를 담을거면 참여하는 post 2개 모두의 member를 알고있어야 하지 않을까요?? 진행중인 대결을 볼때 신청 당한것도 보여줘야 할 것 같아요

thguss reviewed (commented):
수고하셨습니다~! 궁금한 것이 추가로 있는데요!

- ApplicationTests 파일을 삭제한 이유가 있을까요??
- 전체적으로 createdAt, updatedAt 필드가 필요할 것 같습니다! BaseEntity를 이용하면 좋을 것 같아요~!

park-se-jun commented on src/main/java/com/example/demo/model/Music.java:
```diff
@@ -0,0 +1,41 @@
+@NoArgsConstructor
+public class Music {
+
+	// TODO : Music을 따로 엔티티로 두는 것이 맞을지 -> 이렇게 되면 음악 검색 API 하기 이전에 쿼리를 던져야함.
```
api 에 따라 다를듯

park-se-jun commented on src/main/java/com/example/demo/model/Post.java:
```diff
@@ -0,0 +1,41 @@
+	@ManyToOne(fetch = FetchType.LAZY)
+	Music music;
+
+	@OneToMany(mappedBy = "ownerPost")
```
ownerBattles와 oppositbattles  는 뭐가 다른거낙요?

twotwobread committed:
[FIX] Battle 엔티티 수정.
- Post와 득표 수에 대한 필드를 임베디드 타입으로 설정.
- Member 필드 제거.

twotwobread committed:
[ADD] BaseEntity 추가.
- createdAt, updatedAt을 가진 BaseEntity를 이용하여 모든 엔티티에 생성, 수정 시간 추가.

twotwobread committed:
[FIX] Member 엔티티 수정.
- Image 엔티티 제거 -> Member 엔티티에 imageUrl로 추가.
- 필드명 일부 수정.
- 남은 대결권 필드 추가 -> remainingChallenges;
- 소셜 로그인과 관련된 필드 값 추가 -> refreshToken, social
- likes 리스트 필드 추가.

twotwobread committed:
[FIX] Music 관련 코드 수정.
- Music 엔티티 -> 임베디드 타입 클래스로 수정.
- post 관련 연관관계 삭제.

twotwobread committed:
[FIX] Post 엔티티 수정
- 필드 값 일부 수정.
- likes 리스트 필드 추가.

twotwobread committed:
[ADD] Like 엔티티 추가

twotwobread committed:
[ADD] Vote 엔티티 추가.

thguss commented on src/main/java/com/example/demo/model/battle/Battle.java:
```diff
@@ -0,0 +1,28 @@
+	private BattleInfo challengedPost;
+
+	@Embedded
+	private BattleInfo challengingPost;
```
칼럼명 중복으로 인해서 override가 필요할 것 같아요~!

리마인드 겸 다시 언급드려요 ㅎㅅㅎ

thguss commented on src/main/java/com/example/demo/model/battle/Vote.java:
```diff
@@ -0,0 +1,38 @@
+
+	@ManyToOne(fetch = FetchType.LAZY)
+	@JoinColumn(name = "id")
+	private Post post;
```
필드명을 selectedPost 라고 해주면 더 직관적일 것 같아요!

thguss commented on src/main/java/com/example/demo/model/member/Member.java:
```diff
@@ -0,0 +1,55 @@
+	private Long id;
+
+	@NotNull
+	private String imageUrl;
```
개인적으로 profileImageUrl 변수명이 더 직관적일 것 같은데 어떤가요?!?

thguss commented on src/main/java/com/example/demo/model/member/Member.java:
```diff
@@ -0,0 +1,55 @@
+	@NotNull
+	private String nickname;
+
+	private int remainingChallenges;
```
변수명이 좀 더 직관적이었으면 좋겠어요! 예를 들면 challengeTicket !?

park-se-jun replied:
countOfChallengeTicket 같은 느낌은 어떤가요?

thguss commented on src/main/java/com/example/demo/model/member/MemberScore.java:
```diff
@@ -0,0 +1,13 @@
+public class MemberScore {
+	private int ranking;
+	private int vitoryPoint;
+	private int vitoryCount;
```
승리 맞나용?!? victory로 수정 부탁드려요~!

thguss commented on src/main/java/com/example/demo/model/post/Music.java:
```diff
@@ -0,0 +1,28 @@
+
+@Getter
+@Embeddable
+public class Music {
```
String musicImageUrl 도 추가 부탁드려요~!

park-se-jun replied:
musicImageUrl 보단 albumCoverUrl로 하는게 좋을것 같아요

thguss reviewed (commented):
수고하셨습니다!! 언제나 감사드립니다~!

park-se-jun commented on src/main/java/com/example/demo/model/BaseEntity.java:
```diff
@@ -0,0 +1,27 @@
+@Getter
+@MappedSuperclass
+@EntityListeners(AuditingEntityListener.class)
+public abstract class BaseEntity {
```
베이스 엔티티 좋아용 👍

park-se-jun commented on src/main/java/com/example/demo/model/post/Like.java:
```diff
@@ -0,0 +1,33 @@
+	private Long id;
+
+	@ManyToOne(fetch = FetchType.LAZY)
+	@JoinColumn(name = "id")
```
JoinColumn은 FK의 컬럼 명을 말하는걸로 알고 있는데 id면 PK 랑 같아지니까 post에 대해서는 post_id, member에 대해서는 member_id로 변경하는게 맞는것 같아요.



p.s id를 그냥 id 로 해도 괜찮다고 얘기를 했던 부분은 FK 가 아닌 스스로의 id(PK) 에서 어차피 자기꺼니까 id로 해도 무방하다 란 의미로 말했었습니당

park-se-jun commented on src/main/java/com/example/demo/model/member/Member.java:
```diff
@@ -0,0 +1,55 @@
+	private String refreshToken;
+
+	@Embedded
+	private SocialInfo social;
```
그냥 social이면 Social에 대한 것으로 혼동 될수 있을 것 같아 socialInfo로 하는게 좋지 않을까요?

park-se-jun commented on src/main/java/com/example/demo/model/post/Post.java:
```diff
@@ -0,0 +1,58 @@
+	@JoinColumn(name = "id")
+	private Member member;
+
+	@OneToMany(mappedBy = "post")
```
그냥 like와 양방향 연관관계를 맺고 있기 보다는 like count를 가지고 있고 ilke 가 증가할 떄 마다 늘려주거나, like를 보여주는 거는 like 에서 post id로 count 하는 쿼리로 하는 편이 좋을 것.... 아닌가? 일단 이대로 해봐도 괜찮을 것 같기두 하고 잘 모르겠네요 어떻게 생각하나요

thguss replied:
저도 세준님과 비슷한 의견입니다.
MVP 단계에서는 좋아요 리스트는 아직 없으니까, likeCnt만으로 기능 구현이 충분하다고 생각합니다.
좋아요는 likeCnt로 보여주되, List<Like>는 Post에서 가지고 있지 않아도 될 것 같네요!

park-se-jun commented on src/main/java/com/example/demo/model/battle/Vote.java:
```diff
@@ -0,0 +1,38 @@
+	private Long id;
+
+	@ManyToOne(fetch = FetchType.LAZY)
+	@JoinColumn(name = "id")
```
요것들도 다 마찬가지 로 battle_id, post_id, member_id 가 되야하는 것 같습니다

park-se-jun commented on src/main/java/com/example/demo/model/member/Member.java:
```diff
@@ -0,0 +1,55 @@
+	@OneToMany
+	private List<Battle> battles = new ArrayList<>();
+
+	@OneToMany
```
member와 post 간의 양방향 연관관계 설정에서 mappedBy 가 없는 것 같아요

park-se-jun reviewed (commented):
짱수영님 고생 너무 많으십니다........

twotwobread committed:
[FIX] Battle 엔티티 수정.
- AttributeOverride를 이용한 컬럼명 오버라이드.
- AssociationOverride를 이용한 fk 오버라이드.

twotwobread committed:
[FIX] Like 엔티티 수정
- fk명 수정.

twotwobread committed:
[FIX] Member 엔티티 수정.
- 필드명 수정.
- fk명 수정.

twotwobread committed:
[FIX] MemberScore 클래스 수정.
- 필드명 수정.

twotwobread committed:
[FIX] Music 엔티티 수정.
- 필드명 수정.

twotwobread committed:
[FIX] Post 엔티티 수정.
- fk명 수정.
- 필드명 수정.

twotwobread committed:
[FIX] Vote 엔티티 수정.
- fk명 수정.

twotwobread committed:
[CHORE] guava 의존성 추가.

twotwobread committed:
[ADD] Battle 및 BattleInfo에 대한 validation 로직 추가.
- Bean Validation 추가.
- Guava를 통한 application validation 추가.
- 검증 로직
  - 득표 수가 음수가 되면 안됨.

twotwobread committed:
[ADD] Post, Music 및 Genre에 대한 validation 추가.
- Bean Validation 추가.
- Guava를 이용한 application validation 추가.
- 검증 로직
  - Post
    - 좋아요 개수가 음수이면 안됨.
  - Music
    - 음악 클래스는 모든 값이 무조건 필요하기 때문에 Builder를 이용하지 않았음. 근데 JPA 자체에서 기본 생성자를 이용하기에 어노테이션을 달아줌.
    - 음악 고유 번호가 Null이거나 공백으로 이루어지면 안됨.
    - 앨범 커버 이미지 URL이 Null or 공백이면 안되고 최대 2000자 까지
    - 가수 이름이 Null이거나 공백으로 이루어지면 안됨.
    - 음악 제목이 Null이거나 공백으로 이루어지면 안됨.
    - 장르가 Null이면 안됨.
    - 음악 URL이 Null or 공백이면 안되고 최대 2000자 까지

twotwobread committed:
[ADD] Member, MemberScore 및 SocialInfo, Social validation 추가.
- Bean Validation 추가.
- Guava를 이용한 application validation 추가.
- 검증 로직
  - Member
    - 프로필 이미지 URL이 Null or 공백이면 안되고 최대 2000자 까지
    - 닉네임이 Null or 공백이면 안되고 최소 1자부터 24자 까지
    - 대결권 개수가 음수일 수 없음.
    - refresh token이 Null or 공백이면 안됨.
    - 여기도 무조건 필요한 값이 있어서 Builder 말고 생성자를 이용하는게 더 좋아보이긴 함.
  - MemberScore
    - 기본 생성자를 이용하면 0으로 초기화될 것이고 ranking이 들어올수도 있고 아닐수도 있으니까 setter를 이용함.
    - 모든 필드값이 음수만 아니면 됨.
  - SocialInfo
    - 소셜 로그인 타입이 존재해야함.
    - 소셜 ID가 Null이거나 공백이면 안됨.

thguss commented on build.gradle:
```diff
@@ -10,7 +10,7 @@ plugins {
 
 jacoco {
-	toolVersion = '0.8.8'
+    toolVersion = '0.8.8'
```
체크스타일로 맞춰진 사항일까요?!

thguss commented on src/main/java/com/example/demo/model/member/Member.java:
```diff
@@ -0,0 +1,95 @@
+		checkArgument(profileImageUrl.length() <= 2000, "프로필 이미지 URL이 2000자보다 더 길 수 없습니다.", profileImageUrl);
+
+		checkArgument(Objects.nonNull(nickname), "닉네임이 Null일 수 없습니다.", nickname);
+		checkArgument(!nickname.isBlank(), "닉네임이 공백일 수 없습니다.", nickname);
```
isBlank 만으로 nonNull 까지 잡을 수 있는 것으로 알고 있는데, 그게 맞다면 isBlank 로만 유효성 체크해도 될 것 같아요~!

thguss commented on src/main/java/com/example/demo/model/member/Member.java:
```diff
@@ -0,0 +1,95 @@
+
+		checkArgument(Objects.nonNull(nickname), "닉네임이 Null일 수 없습니다.", nickname);
+		checkArgument(!nickname.isBlank(), "닉네임이 공백일 수 없습니다.", nickname);
+		checkArgument(1 <= nickname.length() && nickname.length() <= 24, "닉네임의 길이는 1보다 짧거나 24보다 더 길 수 없습니다.", nickname);
```
1보다 짧으면 공백이고, 이는 isBlank 체크로 충분한 것 같아요!

thguss commented on src/main/java/com/example/demo/model/member/Member.java:
```diff
@@ -0,0 +1,95 @@
+		checkArgument(1 <= nickname.length() && nickname.length() <= 24, "닉네임의 길이는 1보다 짧거나 24보다 더 길 수 없습니다.", nickname);
+
+		checkArgument(Objects.nonNull(refreshToken), "refresh token이 Null일 수 없습니다.", refreshToken);
+		checkArgument(!refreshToken.isBlank(), "refresh token이 공백일 수 없습니다.", refreshToken);
```
이 또한 isBlank 가 nonNull 의 역할까지 가지면 isBlank 로 충분할 것 같아요~

thguss commented on src/main/java/com/example/demo/model/member/Member.java:
```diff
@@ -0,0 +1,95 @@
+		this.countOfChallengeTicket = countOfChallengeTicket;
+		memberScore.setRanking(ranking);
+		memberScore.setVictoryPoint(victoryPoint);
+		memberScore.setVictoryCount(victoryCount);
```
memberScore 를 setter 로 넘겨주는 것보다는, 

memberScore = new memberScore(ranking, victoryPoint, victoryCount);
또는
memberScore.update(ranking, victoryPoint, victoryCount);

위와 같이 한번에 값을 세팅해주는 것이 좋은 방향으로 알고 있어요!

thguss commented on src/main/java/com/example/demo/model/member/Member.java:
```diff
@@ -0,0 +1,95 @@
+		memberScore.setVictoryPoint(victoryPoint);
+		memberScore.setVictoryCount(victoryCount);
+		this.refreshToken = refreshToken;
+		this.socialInfo = new SocialInfo(socialString, socialId);
```
오 따따봉

thguss commented on src/main/java/com/example/demo/model/member/MemberScore.java:
```diff
@@ -0,0 +1,40 @@
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+class MemberScore {
+
+	@Min(value = 0)
```
요 의미는 최솟값이 0이라는 의미인가요?-? (단순 궁금증)

thguss commented on src/main/java/com/example/demo/model/member/MemberScore.java:
```diff
@@ -0,0 +1,40 @@
+		this.victoryPoint = victoryCount;
+	}
+
+	protected void setVictoryCount(int victoryCount) {
```
(위에서 언급한 사항) setter로 만들기 보다는 별도의 생성자 또는 update 메소드를 생성하는 방향이 좋아보입니다.
개인적으로 update 메소드 선호합니다 ㅎㅅㅎ

thguss commented on src/main/java/com/example/demo/model/member/Social.java:
```diff
@@ -0,0 +1,12 @@
+
+	public static boolean of(String socialString) {
+		return Arrays.stream(Social.values())
+			.anyMatch(social -> social.name().equals(socialString));
```
어떤 역할을 하고, 어떤 용도의 메소드인가요?_? (단순 궁금증)

thguss commented on src/main/java/com/example/demo/model/member/SocialInfo.java:
```diff
@@ -0,0 +1,39 @@
+		checkArgument(Social.of(socialString), "해당 소셜 로그인 방식이 존재하지 않습니다.", socialString);
+
+		checkArgument(Objects.nonNull(socialId), "소셜 ID가 Null일 수 없습니다.", socialId);
+		checkArgument(socialId.isBlank(), "소셜 ID가 공백일 수 없습니다.", socialId);
```
여기도 isBlank 로 두 역할을 모두 할 수 있으면, isBlank만 남겨도 좋을 것 같아요!

thguss commented on src/main/java/com/example/demo/model/member/SocialInfo.java:
```diff
@@ -0,0 +1,39 @@
+	@NotBlank
+	private String socialId;
+
+	SocialInfo(String socialString, String socialId) {
```
SocialInfo에는 protected를 붙이지 않은 이유가 있을까요?-?

thguss commented on src/main/java/com/example/demo/model/member/SocialInfo.java:
```diff
@@ -0,0 +1,39 @@
+		checkArgument(Objects.nonNull(socialId), "소셜 ID가 Null일 수 없습니다.", socialId);
+		checkArgument(socialId.isBlank(), "소셜 ID가 공백일 수 없습니다.", socialId);
+
+		this.socialType = Social.valueOf(socialString);
```
Socail.GOOGLE 이 아닌 valueOf로 할당한 이유가 궁금해요! (더불어 생성자 파라미터에서 SocialType이 아닌 String 형식으로 받은 이유도 궁금합니다)

thguss commented on src/main/java/com/example/demo/model/post/Music.java:
```diff
@@ -0,0 +1,78 @@
+
+		// TODO: Genre에 대한 validationd을 어떤 방식으로 해야할지 잘모르겠음.
+		//  -> Genre는 value("DANCE"), name("댄스")
+		//  두 가지 값이 있는데 이 두 가지 값을 넣은 이유가 받을 땐 name으로 받고 저장 시 value를 이용하겠다는 의미 라고 생각함.
```
저는 name으로 받는 것보다는, 응답 데이터로 뿌려줄 때 (value보다 name이 더 알아보기 쉬우니까) name 으로 뿌려주기 위함이라고 생각하고 있었어요!

thguss replied:
하지만 게시물을 등록할 때 장르명으로 genre를 찾는 로직에는 name이 사용되겠네요!

thguss commented on src/main/java/com/example/demo/model/post/Music.java:
```diff
@@ -0,0 +1,78 @@
+		//  두 가지 값이 있는데 이 두 가지 값을 넣은 이유가 받을 땐 name으로 받고 저장 시 value를 이용하겠다는 의미 라고 생각함.
+		//  -> 근데 Music이 만들어질 수 있는 경우가 두 가지임 DB에서 꺼낼 때랑 게시물 등록할 떄
+		//  그래서 validation check할 분기가 2번인데
+		//  생성자에서 어떨 땐 name으로 찾는 메소드로 check, 어떨 땐 value로 찾는 메소드로 check 이 기준을 못나누겠음.
```
체크 분기가 2개로 나눠지는 것이 고민이라면, 생성자를 2개로 나누는 것도 방법이라고 생각합니다.
한 쪽에는 name으로, 다른 한 쪽은 value로 파라미터가 들어오는 경우가 될 것 같아요!

thguss commented on src/main/java/com/example/demo/model/post/Music.java:
```diff
@@ -0,0 +1,78 @@
+		//  생성자에서 어떨 땐 name으로 찾는 메소드로 check, 어떨 땐 value로 찾는 메소드로 check 이 기준을 못나누겠음.
+
+		checkArgument(Objects.nonNull(id), "음악 고유 번호는 Null일 수 없습니다.", id);
+		checkArgument(!id.isBlank(), "음악 고유 번호는 공백일 수 없습니다.", id);
```
여기도 isBlank 만으로 충분할까요??

thguss commented on src/main/java/com/example/demo/model/post/Music.java:
```diff
@@ -0,0 +1,78 @@
+		checkArgument(!id.isBlank(), "음악 고유 번호는 공백일 수 없습니다.", id);
+
+		checkArgument(Objects.nonNull(albumCoverUrl), "앨범 커버 이미지 URL은 Null일 수 없습니다.", albumCoverUrl);
+		checkArgument(!albumCoverUrl.isBlank(), "앨범 커버 이미지 URL은 공백일 수 없습니다.", albumCoverUrl);
```
isBlank !

thguss commented on src/main/java/com/example/demo/model/post/Music.java:
```diff
@@ -0,0 +1,78 @@
+		checkArgument(albumCoverUrl.length() <= 2000, "앨범 커버 이미지 URL은 2000자보다 더 길 수 없습니다.", albumCoverUrl);
+
+		checkArgument(Objects.nonNull(singer), "가수 이름은 Null일 수 없습니다.", singer);
+		checkArgument(!singer.isBlank(), "가수 이름은 공백일 수 없습니다.", singer);
```
isBlank !

thguss commented on src/main/java/com/example/demo/model/post/Music.java:
```diff
@@ -0,0 +1,78 @@
+		checkArgument(!singer.isBlank(), "가수 이름은 공백일 수 없습니다.", singer);
+
+		checkArgument(Objects.nonNull(title), "음악 제목은 Null일 수 없습니다.", title);
+		checkArgument(!title.isBlank(), "음악 제목은 공백일 수 없습니다.", title);
```
isBlank !

thguss commented on src/main/java/com/example/demo/model/post/Music.java:
```diff
@@ -0,0 +1,78 @@
+		checkArgument(!title.isBlank(), "음악 제목은 공백일 수 없습니다.", title);
+
+		checkArgument(Objects.nonNull(musicUrl), "음악 URL은 Null일 수 없습니다.", musicUrl);
+		checkArgument(!musicUrl.isBlank(), "음악 URL은 공백일 수 없습니다.", musicUrl);
```
isBlank !

thguss reviewed (commented):
수고하셨습니다~! 조금만 더 화이팅..!!

park-se-jun commented on src/main/java/com/example/demo/model/post/Post.java:
```diff
@@ -0,0 +1,58 @@
+	private Long id;
+
+	@Embedded
+	private Music music;
```
music 의 id랑 post 의 id 가 겹쳐서 오류가 납니다. 다시 생각해보니 music을 다시 entity로 해서 테이블로 관리해도 괜찮을 것 같아요

park-se-jun commented on src/main/java/com/example/demo/model/post/Post.java:
```diff
@@ -0,0 +1,58 @@
+	private List<Like> likes;
+
+	@OneToMany(mappedBy = "challengedPost")
+	private List<Battle> challengedBattles = new ArrayList<>();
```
컴파일 시 
```
Failed to initialize JPA EntityManagerFactory: [PersistenceUnit: default] Unable to build Hibernate SessionFactory; nested exception is org.hibernate.MappingException: collection foreign key mapping has wrong number of columns: com.example.demo.model.post.Post.challengedBattles type: long
```
요런 오류가 나는데 해결 방법은 못찾았어요😢

park-se-jun reviewed (commented):
슬슬끝나가는거 ㄱㅌ아서 컴ㄴ파일해보고 있는데 오류를 못 해결했습니다.. 같이 얘기해 봅시당

thguss committed:
[FIX] 최소 길이 제한 기준 변경

thguss committed:
[FIX] 회원 스코어 관련 업데이트 메소드 로직 추가 및 반영

thguss committed:
[FIX] String to Enum 타입으로 변경

thguss committed:
[DEL] 미사용 메소드 삭제

thguss committed:
[FIX] Music id 필드명 수정

thguss committed:
[FIX] 접근 지정자 변경

thguss committed:
[TEST] h2 테스트

thguss committed:
[FIX] 테이블명 변경 (예약어 이슈)

thguss committed:
Merge pull request #49 from prgrms-web-devcourse/feature_#48

[CHORE] Entity 설계 의논 및 반영

thguss committed:
[FIX] 그래들 버전 업그레이드

thguss committed:
[FIX] jacoco 활성화 off

thguss committed:
[FIX] jacoco 활성화 on

thguss committed:
[FIX] jacoco 활성화 off

thguss committed:
[ADD] copySwaggerUI 의존성 추가

thguss reviewed (approved):
수고하셨습니다,, 모두 짱

park-se-jun reviewed (approved):
👍👍👍👍
